source/agora/flash/Node.d

```diff
@@ -834,12 +893,29 @@ public abstract class FlashNode : ControlFlashAPI
             this.shared_secrets.remove(payment_hash);
             this.payment_path.remove(payment_hash);
             this.payment_errors.remove(payment_hash);
+
+            // get the invoice if it exists and not just the pointer (GC safety)
+            Invoice inv;
+            auto invoice = this.invoices.get(payment_hash, Invoice.init);
             this.invoices.remove(payment_hash);
+
+            if (invoice != Invoice.init)
+                this.listener.onPaymentSuccess(invoice);
         }
 
         foreach (payment_hash; rev_htlcs)
+        {
             if (auto invoice = payment_hash in this.invoices)
-                this.payInvoice(*invoice); // Retry
+            {
+                ErrorCode error;
+                if (auto errors = payment_hash in this.payment_errors)
+                    error = (*errors)[0].err;  // pick first known reason
```

Comment:
This should  be the last known reason as we queue new failure reasons to the end.
But still, we may report wrong failure reason to the wallet. As error are propagated separately, there is a race between the error message and HTLC being revoked.

Reply:
Yeah I had the same thoughts, not sure what is the best approach. I think I will change it to use the last error then.